Source/Processors/GenericProcessor/GenericProcessor.cpp

```diff
@@ -1177,17 +1177,23 @@ double GenericProcessor::getFirstTimestampForBlock(uint16 streamId) const
     return startTimestampsForBlock.at(streamId);
 }
 
-int64 GenericProcessor::getTimestampSampleIndexForBlock(uint16 streamId) const
-{
-    return timestampSampleIndexForBlock.at(streamId);
+
+std::optional<std::pair<int64, double>> GenericProcessor::getReferenceSampleForBlock(uint16 streamId){
+    try {
+        std::optional<std::pair<int64, double>> sample  = referenceSamplesForBlock.at(streamId);
+        referenceSamplesForBlock.at(streamId) = std::nullopt;
+        return sample;
```

Comment:
@admunkucsd it seems a bit odd to also clear the reference sample here. This means if a processor only intermittently calls `getReferenceSampleForBlock` for whatever reason, then there's a chance to have stale information from a previous block. How about instead we add a `clearReferenceSample` (or allow `setReferenceSample` to take an optional) that clears the `referenceSamplesForBlock` and have `SourceNode` and `FileReader` call this? Makes it much easier to follow this logic too.

Reply:
I see the issue here, but I want to factor in how blocks are created through `addToBuffer`/`readAllFromBuffer`. Since there's no guarantee that the blocks created by `addToBuffer` will be organized the same as Blocks that are passed through the `AudioBuffer`, then we have to have some way to associate when a `ReferenceSample` was added relative to which samples were added to the `DataBuffer` at the same time. 

The current solution I have here is store the last `ReferenceSample` taken alongside each sample, but the main issue is that stale data is in fact being added with every block and the Synchronizer manages only using new data in the linear regressions.  I think the intention of both of the suggested solutions is to only have the` ReferenceSample` event added once per ReferenceSample, which is much easier to follow. 

However, there is still the issue with aligning with continuous data samples. What I could do is I can have `SourceNode`/`FileReader` save the last ReferenceSample taken and only add events when a different ReferenceSample is added to a Block

Reply:
Actually, I think the solutions you proposed do exactly this, I just misinterpreted what the implementation would look like